src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp

```diff
@@ -432,92 +440,124 @@ gsl::span<const winrt::com_ptr<Profile>> SettingsLoader::_getNonUserOriginProfil
+void SettingsLoader::_parse(const OriginTag origin, const winrt::hstring& source, const std::string_view& content, ParsedSettings& settings)
 {
-    const auto json = content.empty() ? Json::Value{ Json::ValueType::objectValue } : _parseJSON(content);
-    const auto& profilesObject = _getJSONValue(json, ProfilesKey);
-    const auto& defaultsObject = _getJSONValue(profilesObject, DefaultSettingsKey);
-    const auto& profilesArray = profilesObject.isArray() ? profilesObject : _getJSONValue(profilesObject, ProfilesListKey);
+    const auto json = _parseJson(content);
+
+    settings.clear();
 
-    // globals
     {
-        settings.globals = GlobalAppSettings::FromJson(json);
+        settings.globals = GlobalAppSettings::FromJson(json.root);
 
-        if (const auto& schemes = _getJSONValue(json, SchemesKey))
+        for (const auto& schemeJson : json.colorSchemes)
         {
-            for (const auto& schemeJson : schemes)
+            if (const auto scheme = ColorScheme::FromJson(schemeJson))
             {
-                if (schemeJson.isObject())
-                {
-                    if (const auto scheme = ColorScheme::FromJson(schemeJson))
-                    {
-                        settings.globals->AddColorScheme(*scheme);
-                    }
-                }
+                settings.globals->AddColorScheme(*scheme);
             }
         }
     }
 
-    // profiles.defaults
     {
-        settings.baseLayerProfile = Profile::FromJson(defaultsObject);
+        settings.baseLayerProfile = Profile::FromJson(json.profileDefaults);
         // Remove the `guid` member from the default settings.
         // That will hyper-explode, so just don't let them do that.
         settings.baseLayerProfile->ClearGuid();
         settings.baseLayerProfile->Origin(OriginTag::ProfilesDefaults);
     }
 
-    // profiles.list
     {
-        const auto size = profilesArray.size();
-
-        // NOTE: This function is supposed to *replace* the contents of ParsedSettings. Don't break this promise.
-        // SettingsLoader::FindFragmentsAndMergeIntoUserSettings relies on this.
-        settings.profiles.clear();
+        const auto size = json.profilesList.size();
         settings.profiles.reserve(size);
-
-        settings.profilesByGuid.clear();
         settings.profilesByGuid.reserve(size);
 
-        for (const auto& profileJson : profilesArray)
+        for (const auto& profileJson : json.profilesList)
         {
-            auto profile = Profile::FromJson(profileJson);
-            profile->Origin(origin);
-
-            // The Guid() generation below depends on the value of Source().
-            // --> Provide one if we got one.
-            if (!source.empty())
+            auto profile = _parseProfile(origin, source, profileJson);
+            if (profile->HasGuid())
             {
-                profile->Source(source);
+                _appendProfile(std::move(profile), settings);
             }

```

Comment:
Won't this always be `true` because `_parseProfile` generates/saves a guid if there isn't one?

Reply:
It'll only generate a Guid() if there was a Name().
Normally I'd just return a nullptr to communicate to the caller that no profile could be parsed, but fragments allow profiles without either Guid or Name. They also accept profiles with a valid Updates() value. That's why both callers check HasGuid(). I'll add comments.